Applications/spire/source/charting/cached_chart_model.cpp

```diff
@@ -0,0 +1,89 @@
+#include "spire/charting/cached_chart_model.hpp"
+
+using namespace boost::signals2;
+using namespace Spire;
+
+CachedChartModel::CachedChartModel(ChartModel& model)
+    : m_chart_model(&model) {}
+
+ChartValue::Type CachedChartModel::get_x_axis_type() const {
+  return m_chart_model->get_x_axis_type();
+}
+
+ChartValue::Type CachedChartModel::get_y_axis_type() const {
+  return m_chart_model->get_y_axis_type();
+}
+
+QtPromise<std::vector<Candlestick>> CachedChartModel::load(ChartValue first,
+    ChartValue last) {
+  for(auto& range : m_ranges) {
```

Comment:
Perform a binary search, linear search is too slow.

Reply:
Will I have to implement this myself? I haven't found anything suitable in the standard library.